### PR TITLE
[Application] add object detection example using pytorch

### PR DESCRIPTION
**Pull request by baek2sm**
Add YOLO v2 object detection example using pytorch.

It was implemented in the same way as the YOLO v2 model, but only the backbone model was made into a simpler cnn model to make it easier to configure the initial version of YOLO to be implemented by NNTrainer.

In order to clarify the parts to be implemented by NNTrainer, it is a code that constructs only the basic structure proposed in the original yolo paper, excluding several third-part training techniques.

### CODE REVIEW

**Review by DonghakPark — APPROVED** [review]
LGTM!!
But There was doxygen fail, please check

**Comment by baek2sm** [comment]
> LGTM!! But There was doxygen fail, please check

Thanks, I'll update annotation of these codes. :)

File: Applications/YOLO/PyTorch/main.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        (11.2364, 10.0071)]
+    )
+
+    outsize = (13, 13)
+    width, height = outsize
+    
+    # restore cell pos to x, y    
+    for w in range(width):
+        for h in range(height):
+            bbox_pred[0, height*h + w, :, 0] += w
```
Comment (lhs8928) [inline Applications/YOLO/PyTorch/main.py:101]: I guess this code will only process first batch isn't it?

Comment (baek2sm) [reply Applications/YOLO/PyTorch/main.py:101]: You are right. thanks! I'll fix it.

File: Applications/YOLO/PyTorch/yolo_loss.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+            (11.2364, 10.0071)]
+        )
+                
+        self.mse = nn.MSELoss()
+        self.bbox_loss, self.iou_loss, self.cls_loss = None, None, None       
+        
+    def forward(self, bbox_pred, iou_pred, prob_pred, bbox_gt, cls_gt):
+        """
+        param bbox_pred: shape(batch_size, cell_h x cell_w, num_anchors, 4)
+        param iou_pred: shape(batch_size, 1)
```
[the remark this answers is not in the document]
Comment (baek2sm) [reply Applications/YOLO/PyTorch/yolo_loss.py:69]: It should be (batch_size, cell_h x cell_w, num_anchors, 1). thanks.

File: Applications/YOLO/PyTorch/yolo_loss.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        param prob_pred: shape(batch_size, cell_h x cell_w, num_anchors, num_classes)
+        param bbox_gt: shape(batch_size, num_bbox, 4), data range(0~1)
+        param cls_gt: shape(batch_size, 1)
```
[the remark this answers is not in the document]
Comment (baek2sm) [reply Applications/YOLO/PyTorch/yolo_loss.py:72]: That's right. thanks!

**Review by lhs8928 — APPROVED** [review]
LGTM

File: Applications/YOLO/PyTorch/dataset.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        self.input_images = np.array(self.input_images)
+        self.input_images = torch.FloatTensor(self.input_images).permute((0, 3, 1, 2))
+        
+    def __len__(self):
+        return self.length
+    
+    def __getitem__(self, idx):
+        return self.input_images[idx], self.bbox_gt[idx], self.cls_gt[idx]
+    
+def collate_db(batch):
```
Comment (DonghakPark) [inline Applications/YOLO/PyTorch/dataset.py:44]: For CI PASS in  ```Applications/YOLO/PyTorch/dataset.py``` 
you need function doxygen tag 
For example)
```.py
"""
@param batch batch size
@return items items
"""
```

File: Applications/YOLO/PyTorch/yolo_loss.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+        _cls_mask = torch.zeros((hw, num_anchors, 1))
+                        
+        # find best anchors
+        _bbox_gt_wh = _bbox_gt.clone()[:, 2:]        
+        best_anchors = find_best_ratio(self.anchors, _bbox_gt_wh)
+        
+        # normalize x, y pos based on cell coornindates
+        cell_w = float(self.img_shape[0]) / self.outsize[0] 
+        cell_h = float(self.img_shape[1]) / self.outsize[1]
+        cx = _bbox_gt[:, 0] * self.img_shape[0] / cell_w
```
Comment (lhs8928) [inline Applications/YOLO/PyTorch/yolo_loss.py:180]: How about this?
`cx = _bbox_gt[:, 0] * self.outsize[0]`
I think cell_w, cell_h does not needed

Comment (baek2sm) [reply Applications/YOLO/PyTorch/yolo_loss.py:180]: > How about this? `cx = _bbox_gt[:, 0] * self.outsize[0]` I think cell_w, cell_h does not needed

That's a good idea. thanks!

File: Applications/YOLO/PyTorch/dataset.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
+##
```
Comment (jijoongmoon) [inline Applications/YOLO/PyTorch/dataset.py:1]: please add SPDX-License-Identifier: Apache-2.0 
as in https://github.com/nnstreamer/nntrainer/blob/6bf3610820817d298185d9dce8eda9d84a3a71ac/Applications/MNIST/PyTorch/main.py#L2

File: Applications/YOLO/PyTorch/main.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
+##
```
Comment (jijoongmoon) [inline Applications/YOLO/PyTorch/main.py:1]: same here

File: Applications/YOLO/PyTorch/main.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+# load data
+dataset = YOLODataset(img_dir, ann_dir)
+loader = DataLoader(dataset, batch_size=batch_size, collate_fn=collate_db, shuffle=True, drop_last=True)
+
+
+# set model, loss and optimizer
+model = YoloV2_light(num_classes=5)
+criterion = YoloV2_LOSS(num_classes=5)
+optimizer = optim.Adam(model.parameters(), lr=1e-3)
+scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=10, eta_min=0)
```
Comment (jijoongmoon) [inline Applications/YOLO/PyTorch/main.py:43]: If we need to use this lr_scheduler, then we need to implement it into NNTrainer

File: Applications/YOLO/PyTorch/yolo_loss.py [hunk 9]
```diff
@@ -0,0 +1,203 @@
+##
```
Comment (jijoongmoon) [inline Applications/YOLO/PyTorch/yolo_loss.py:1]: same here.

File: Applications/YOLO/PyTorch/yolo_loss.py [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+    b1 = np.divide(anchors[:, 0], anchors[:, 1])
+    b2 = np.divide(bbox[:, 0], bbox[:, 1])
+    similarities = np.abs(b1.reshape(-1, 1) - b2)
+    best_match = np.argmin(similarities, axis=0)
+    return best_match
+
+##
+# @brief loss class for yolo
+class YoloV2_LOSS(nn.Module):
+    """Yolo v2 loss"""
```
Comment (jijoongmoon) [inline Applications/YOLO/PyTorch/yolo_loss.py:63]: when you implement this loss into NNTrainer, plz consider where these computations should be done among calDerivatives, calGradient.

**Review by jijoongmoon — APPROVED** [review]
LGTM! Great Work!

File: Applications/YOLO/PyTorch/yolo_loss.py (hunk 10 above)
Comment (baek2sm) [reply Applications/YOLO/PyTorch/yolo_loss.py:63]: thank you, I will consider your comment and work on it.

File: Applications/YOLO/PyTorch/dataset.py (hunk 6 above)
Comment (baek2sm) [reply Applications/YOLO/PyTorch/dataset.py]: I added this line. thanks!

File: Applications/YOLO/PyTorch/main.py (hunk 8 above)
Comment (baek2sm) [reply Applications/YOLO/PyTorch/main.py:43]: actually, It's an optional element. so I think it's okay to add it later. thanks!

File: Applications/YOLO/PyTorch/main.py (hunk 7 above)
Comment (baek2sm) [reply Applications/YOLO/PyTorch/main.py]: thank you!

File: Applications/YOLO/PyTorch/yolo_loss.py (hunk 9 above)
Comment (baek2sm) [reply Applications/YOLO/PyTorch/yolo_loss.py]: thank you
